--- PRoot/src/GNUmakefile       2016-11-28 12:24:12.000000000 +0000
+++ PRoot-static/src/GNUmakefile        2016-12-06 11:35:24.438382099 +0000
@@ -14,11 +14,11 @@
 OBJCOPY  = $(CROSS_COMPILE)objcopy
 OBJDUMP  = $(CROSS_COMPILE)objdump
 
-CPPFLAGS += -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -I. -I$(VPATH)
-CFLAGS   += -Wall -Wextra -O2
-LDFLAGS  += -ltalloc
+CPPFLAGS += -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -I. -I$(VPATH) 
+CFLAGS   += -Wall -Wextra -O2 
+LDFLAGS  += -L. -ltalloc -static 
 
-CARE_LDFLAGS = -larchive
+CARE_LDFLAGS = -larchive -static 
 
 OBJECTS += \
        cli/cli.o               \
